node/core/pvf/src/worker_common.rs

```diff
@@ -0,0 +1,280 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Common logic for implementation of worker processes.
+
+use crate::LOG_TARGET;
+use async_std::{
+	io,
+	os::unix::net::{UnixListener, UnixStream},
+	path::{PathBuf, Path},
+};
+use futures::{
+	AsyncRead, AsyncWrite, AsyncReadExt as _, AsyncWriteExt as _, FutureExt as _, never::Never,
+};
+use futures_timer::Delay;
+use rand::Rng;
+use std::{
+	fmt, mem,
+	pin::Pin,
+	str::FromStr,
+	task::{Context, Poll},
+	time::Duration,
+};
+use pin_project::pin_project;
+
+/// This is publicly exposed only for integration tests.
+#[doc(hidden)]
+pub async fn spawn_with_program_path(
+	debug_id: &'static str,
+	program_path: &str,
+	extra_args: &'static [&'static str],
+	spawn_timeout_secs: u64,
+) -> Result<(IdleWorker, WorkerHandle), SpawnErr> {
+	with_transient_socket_path(debug_id, |socket_path| {
+		let socket_path = socket_path.to_owned();
+		let program_path = PathBuf::from(program_path);
```

Comment:
Why don't we pass this directly as `PathBuf`? 

At the main calling site, convert the path before into some string. Looks weird.

Reply:
Ok, so I changed this, however not directly to `PathBuf` but `impl Into<PathBuf>`. One of the three call sites is actually a string.